### PR TITLE
ec2-room-validation

### PR DESCRIPTION
**Pull request by ms-petrugarstea**
places an instance in a subnet
calculates the available ips

### CODE REVIEW

**Comment by ms-petrugarstea** [comment]
example:
treadmill --dns-domain ipa.poc.cloud.ms.com admin aws instance create --image ms-generic --image-account 876489724231 --secgroup sg-49bf3839 --key SymphonyPOC --size t2.large --subnet subnet-4f3b3b07 --subnet VPC101SNCSR2D --subnet VPC101SNCSR2D --count 15

Explanation:
The cli takes care of subnet duplication.
The availability of the 2 subnets in total is equal to 14 hosts.
In cli I specified 15 hosts, so the cli will spin up the 14 hosts and will raise an exception that in above mentioned subnets has left no space for 1 instance.

File: lib/python/treadmill_aws/cli/admin/aws/instance.py [hunk 1]
```diff
@@ -137,23 +136,42 @@ def create(
...
+                key=key,
+                secgroup_ids=secgroup_id,
+                instance_type=size,
+                subnet_id=network[0]['SubnetId'],
+                role=role,
+                instance_vars=instance_vars,
+                instance_profile=instance_profile
+            )
+
+            for hostname in hostnames:
```
Comment (andreikeis) [inline lib/python/treadmill_aws/cli/admin/aws/instance.py]: dont change the output format - it should be list of hosts as before.

File: lib/python/treadmill_aws/cli/admin/aws/instance.py [hunk 2]
```diff
@@ -137,23 +136,42 @@ def create(
...
         )
-        for hostname in hostnames:
-            click.echo(hostname)
+
+        total_subnets = list(
+            [ec2client.list_subnets(ec2_conn, subnet_id.split()) for subnet_id
+             in subnet_ids]
+        )
+
+        value = 1
```
Comment (andreikeis) [inline lib/python/treadmill_aws/cli/admin/aws/instance.py]: not needed. aside from that, it is not a good variable name. subnet_idx is better (yet this is not needed at all).

File: lib/python/treadmill_aws/hostmanager.py [hunk 3]
```diff
@@ -3,6 +3,7 @@
 import time
 import yaml
 
+from botocore.exceptions import ClientError
```
Comment (andreikeis) [inline lib/python/treadmill_aws/hostmanager.py]: do not import classes, we import modules and use qualified names for classes later on.

from botocore import exceptions as boto_exceptions

...

except botoexceptions.ClientError as err:
...

File: lib/python/treadmill_aws/hostmanager.py [hunk 4]
```diff
@@ -47,19 +48,34 @@ def create_host(ec2_conn, ipa_client, image_id, count, domain,
...
+                subnet_id=subnet_id,
+                instance_profile=instance_profile,
+                disk=disk
+            )
+            hosts.append(host_ctx['hostname'])
+
+        except ClientError as error:
+            error_code = error.response['Error']['Code']
+
+            if error_code == 'InsufficientFreeAddressesInSubnet':
```
Comment (andreikeis) [inline lib/python/treadmill_aws/hostmanager.py]: is there a constant for that maybe in boto?

File: lib/python/treadmill_aws/hostmanager.py [hunk 5]
```diff
@@ -47,19 +48,34 @@ def create_host(ec2_conn, ipa_client, image_id, count, domain,
...
+                try:
+                    ipa_client.unenroll_host(hostname=host_ctx['hostname'])
+                except (KeyError, ipaclient.NotFoundError):
+                    pass
```
Comment (andreikeis) [inline lib/python/treadmill_aws/hostmanager.py]: add logging. 

search for _LOGGER in the other files, see how it is used.

File: lib/python/treadmill_aws/hostmanager.py [hunk 6]
```diff
@@ -47,19 +48,34 @@ def create_host(ec2_conn, ipa_client, image_id, count, domain,
...
+
+                break
+
+            else:
+                raise error
```
Comment (andreikeis) [inline lib/python/treadmill_aws/hostmanager.py]: you should change the docstring for the function, describing new behavior. should also add unittest for that.

File: lib/python/treadmill_aws/hostmanager.py (hunk 4 above)
Comment (ms-petrugarstea) [reply lib/python/treadmill_aws/hostmanager.py]: yes, it is.
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/errors-overview.html

**Pull request closed without merging**